### ISSUE TITLE
Change how to write rules

### ISSUE THREAD

**Issue by fisker** [eslint/eslint#14661]
**The version of ESLint you are using.**

Latest

**The problem you want to solve.**

Simplify rule implement

**Your take on the correct solution to problem.**

N/A

**Are you willing to submit a pull request to implement this change?**

Yes

----

As I understand the `create` function of rule returns an object with `listeners`(also known as "callback functions"), normally, the listener function call `context.report()` or collect information, the returned value are ignored.

My suggestion is to make the listener returns `Problem` or `Problem[]`, ESLint will auto call `context.report()` with them, a rule will like

```js
module.export = {
	create(context) {
		const identifiers = new Set();

		return {
			Identifier(node) {
				// This listener don't report a problem
				identifiers.push(node)
			}
			FunctionExpression(node) {
				// This listener report ONE problem
				return {node, messageId: 'no-function-expression'}
			},
			MemberExpression(node) {
				// This listener use array to report multiple problems
				const problems = [];
				if (foo) {
					problems.push({node, messageId: 'problem-1'})
				}
				if (bar) {
					problems.push({node, messageId: 'problem-2'})
				}
				return problems;
			},
			* 'Program:exit'() {
				// This listener use generator to yield multiple problems
				for (const node of identifiers) {
					yield {node, messageId: 'identifier', data: {name: node.name}};
				}
			}
		}
	},
	// ... other meta
};
```

I think this change will simplify the rule, we can do early return without calling `context.report()` first.

The new listener similar to the `fix` function. It can also return `null` or `Iterable`.

**Comment by nzakas** [comment]
I’m sorry, I don’t see any benefit to making such a change. It doesn’t seem to be addressing any major pain point and only increases the complexity of ESLint, as we will have to maintain multiple ways to report errors since we can’t remove context.report(). 

If you would like to write rules in the manner you describe, you can definitely create your own utility that can take a rule written in that format and convert it to use context.report().

**Comment by fisker** [comment]
Yah, I was going to create my own utility, but worth to take a shot here.
